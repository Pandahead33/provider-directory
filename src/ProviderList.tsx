import React, { ReactElement } from "react";
import { Contact } from "./Contact";
import Fields from "./Fields";
import Action from "./Action";

enum SortDirection {
  ascending = "ascending",
  descending = "descending",
}

interface ProviderListProps {
  contacts: Array<Contact>;
  deleteProvider: (id: string) => void;
  openFormPanel: (actionType: Action, record: Contact | null) => void;
  resetProviderList: () => void; 
}

interface ProviderListState {
  sortBy: string;
  sortDirection: SortDirection;
  showSearchResults: boolean;
  searchResults: Array<Contact>;
  /* eslint-disable */
  // to access fields by using [fieldName], this needs to accept
  // any type- which eslint intentionally warns about
  [key: string]: any;
  /* eslint-enable */
}

class ProviderList extends React.Component<
  ProviderListProps,
  ProviderListState
> {
  constructor(props: ProviderListProps) {
    super(props);
    this.state = {
      sortBy: "lastName",
      sortDirection: SortDirection.ascending,
      searchTerm: "",
      searchResults: [],
      showSearchResults: false,
    };

    this.changeSortByFieldDropdown = this.changeSortByFieldDropdown.bind(this);
    this.updateSearchTerm = this.updateSearchTerm.bind(this);
    this.searchProviderList = this.searchProviderList.bind(this);
    this.setSortDirection = this.setSortDirection.bind(this);
  }

  setSortDirection(event: React.SyntheticEvent): void {
    const target = event.target as HTMLInputElement;
    const sortDirection =
      target.value === "ascending"
        ? SortDirection.ascending
        : SortDirection.descending;

    this.setState(
      {
        sortDirection: sortDirection,
      },
      this.sortByField
    );
  }

  changeSortByFieldDropdown(event: React.SyntheticEvent): void {
    const target = event.target as HTMLInputElement;

    this.setState(
      {
        sortBy: target.value,
      },
      this.sortByField
    );
  }

  updateSearchTerm(event: React.SyntheticEvent): void {
    const target = event.target as HTMLInputElement;
    this.setState({
      searchTerm: target.value,
    });
  }

  searchProviderList(): void {
    let searchResults = [...this.props.contacts];
    const fitsSearchCriteria = (contact: Contact) =>
      Object.values(contact).reduce(
        (boolean: boolean, field) =>
          field.toUpperCase().includes(this.state.searchTerm.toUpperCase()) ||
          boolean,
        false
      );

    searchResults = searchResults.reduce(
      (contacts: Array<Contact>, contact) => {
        if (fitsSearchCriteria(contact)) {
          contacts.push(contact);
        }
        return contacts;
      },
      []
    );

    this.setState(
      {
        searchResults: searchResults,
        showSearchResults: !!this.state.searchTerm,
      },
      this.sortByField
    );
  }

  sortByField(): void {
    const fieldName = this.state.sortBy;
    const providerContacts = this.state.showSearchResults
      ? this.state.searchResults
      : this.props.contacts;
    const ascendingOrder = this.state.sortDirection === SortDirection.ascending;

    providerContacts.sort((a, b) => {
      if (a[fieldName] < b[fieldName]) {
        return ascendingOrder ? -1 : 1;
      } else if (a[fieldName] > b[fieldName]) {
        return ascendingOrder ? 1 : -1;
      }

      return 0;
    });
  }

  render(): ReactElement {
    const cells: Array<ReactElement> = [];
    const data = this.state.showSearchResults
      ? this.state.searchResults
      : this.props.contacts;

    this.sortByField();

    data.forEach((record) => {
      const recordValues = Object.values(record);
      const recordKeys = Object.keys(record);

      for (let i = 1; i < recordValues.length; i++) {
        cells.push(
          <div key={JSON.stringify(record) + " " + i} className="cell">
            <p>
              {recordKeys[i] !== Fields.EMAIL_ADDRESS.toString() ? (
                recordValues[i]
              ) : (
                <a href={`mailto:${recordValues[i]}`}>{recordValues[i]}</a>
              )}
            </p>
          </div>
        );
      }

      cells.push(
        <div key={JSON.stringify(record) + " delete"} className="cell">
          <button
            className="action-button"
            onClick={() => this.props.openFormPanel(Action.edit, record)}
          >
            <span role="img" aria-label="edit-icon">
              üñäÔ∏è
            </span>{" "}
            Edit
          </button>
          <button
            className="action-button"
            onClick={() => this.props.deleteProvider(record.id)}
          >
            <span role="img" aria-label="delete-icon">
              ‚ùå
            </span>{" "}
            Delete
          </button>
        </div>
      );
    });

    let providerDataTable = <p>No providers found.</p>;

    if (data.length > 0) {
      providerDataTable = (
        <div style={{ height: "0" }}>
          <div className="provider-table-heading">
            <div className="table-header">
              <p>First Name</p>
            </div>
            <div className="table-header">
              <p>Last Name</p>
            </div>
            <div className="table-header">
              <p>Email Address</p>
            </div>
            <div className="table-header">
              <p>Speciality</p>
            </div>
            <div className="table-header">
              <p>Practice Name</p>
            </div>
            <div className="table-header">
              <p>Action</p>
            </div>
          </div>
          <div className="search-results-container">{cells}</div>
        </div>
      );
    }

    return (
      <div className="provider-list">
        <div className="search-container">
          <h1>Provider List</h1>
          <input
            className="add-new-provider-button"
            type="button"
            value="‚ûï Add New Provider"
            onClick={() => this.props.openFormPanel(Action.add, null)}
          />
          <input
            className="add-new-provider-button"
            type="button"
            value="‚ò†Ô∏è Delete All Providers"
            onClick={this.props.resetProviderList}
          />
          <div className="search-bar-container">
            <input
              className="search"
              onKeyDown={(event) => {
                if (event.key === "Enter") {
                  this.searchProviderList();
                }
              }}
              onChange={this.updateSearchTerm}
              type="search"
            />
            <input
              type="button"
              value="Search"
              onClick={this.searchProviderList}
            />
            <label htmlFor="sortBy">
              <div className="sort-by-label">Sort By</div>
              <select
                id="sortBy"
                name="sortBy"
                value={this.state.sortBy}
                onChange={this.changeSortByFieldDropdown}
              >
                <option value="firstName">First Name</option>
                <option value="lastName">Last Name</option>
                <option value="emailAddress">Email Address</option>
                <option value="speciality">Speciality</option>
                <option value="practiceName">Practice Name</option>
              </select>
            </label>
            <div className="sort-direction">
              <div>
                <input
                  type="radio"
                  name="sortDirection"
                  id="ascending"
                  value="ascending"
                  onChange={this.setSortDirection}
                  checked={this.state.sortDirection === SortDirection.ascending}
                />
                <label htmlFor="ascending">Ascending</label>
              </div>
              <div>
                <input
                  type="radio"
                  name="sortDirection"
                  id="descending"
                  value="descending"
                  onChange={this.setSortDirection}
                  checked={
                    this.state.sortDirection === SortDirection.descending
                  }
                />
                <label htmlFor="ascending">Descending</label>
              </div>
            </div>
          </div>
        </div>
        <span className="results">{data.length} results</span>
        {providerDataTable}
      </div>
    );
  }
}

export default ProviderList;
