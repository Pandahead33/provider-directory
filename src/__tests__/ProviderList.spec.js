import React, { Component } from "react"
import {render, screen, fireEvent} from '@testing-library/react'
import { create } from "react-test-renderer";
import ProviderList from "../ProviderList";

const multipleValidContacts = [
    {
        id: "ecf54597-f668-48f0-af8f-1a654d80f5ec",
        firstName: "Sleepy",
        lastName: "Jimbo",
        emailAddress: "nick@nite.neutron",
        speciality: "Somnologist",
        practiceName: "Sleepy Hollows Clinic",
    },
    {
        id: "98e5eccc-1f6d-47f3-9882-e57cd93a73f9",
        firstName: "Adam",
        lastName: "Smith",
        emailAddress: "westward@expansion.com",
        speciality: "General Medicine",
        practiceName: "Capital Practitioners Primary Care",
    },
    {
        id: "7494e300-d378-4095-a362-dc452d3f8f27",
        firstName: "Zebra",
        lastName: "Elephantress",
        emailAddress: "safari@joe.tech",
        speciality: "Other",
        practiceName: "Safari Joe Veterinary Institute",
    },
    {
        id: "dbd8f84c-79ac-430f-8463-56057a53158a",
        firstName: "Dan",
        lastName: "Smith",
        emailAddress: "sam@last.name",
        speciality: "Podiatrist",
        practiceName: "Generic Foot Meds",
    },
];

test('render and matches snapshot', () => {
    expect(create(<ProviderList contacts={[]}/>).toJSON()).toMatchSnapshot();
});

test('should display "No providers found." message if zero contacts provider', () => {
    render(<ProviderList contacts={[]}/>);

    expect(screen.getByText("0 results").textContent).toBe("0 results");
    expect(screen.getByText("No providers found.").textContent).toBe("No providers found.");
});

test("should display all provided contacts", () => {
    render(<ProviderList contacts={multipleValidContacts}/>);
    expect(screen.getByText(`${multipleValidContacts.length} results`).textContent).toBe(`${multipleValidContacts.length} results`);
});

test("should sort list by any of fields in ascending order by lastName", () => {
    const {container} = render(<ProviderList contacts={multipleValidContacts}/>);
    const tableContent = container.firstChild.childNodes[2].childNodes[1].textContent;

    const descendingRadioButton = screen.getByLabelText('Descending');
    fireEvent.click(descendingRadioButton);
    
    expect(descendingRadioButton.value).toBe("descending");
    expect(descendingRadioButton.checked).toEqual(true);
    
    const ascendingRadioButton = screen.getByLabelText('Ascending');
    fireEvent.click(ascendingRadioButton);

    const firstContactIndex = tableContent.indexOf("Elephantress");
    const secondContactIndex = tableContent.indexOf("Jimbo");
    const thirdContactIndex = tableContent.indexOf("Smith");

    expect(firstContactIndex).toBeLessThan(secondContactIndex);
    expect(secondContactIndex).toBeLessThan(thirdContactIndex);
});

test("should sort list by any of fields in descending order", () => {
    const {container} = render(<ProviderList contacts={multipleValidContacts}/>);

    const descendingRadioButton = screen.getByLabelText('Descending');
    fireEvent.click(descendingRadioButton);
    
    expect(descendingRadioButton.value).toBe("descending");
    expect(descendingRadioButton.checked).toEqual(true);
    
    const tableContent = container.firstChild.childNodes[2].childNodes[1].textContent;

    const firstContactIndex = tableContent.indexOf("Elephantress");
    const secondContactIndex = tableContent.indexOf("Jimbo");
    const thirdContactIndex = tableContent.indexOf("Smith");

    expect(firstContactIndex).toBeGreaterThan(secondContactIndex);
    expect(secondContactIndex).toBeGreaterThan(thirdContactIndex);
});

test("keyword search should filter", () => {
    const {container} = render(<ProviderList contacts={multipleValidContacts}/>);

    const searchBar = screen.getByPlaceholderText('Search...');
    fireEvent.change(searchBar, { target: { value: "westward"}});
    
    const tableContent = container.firstChild.childNodes[2].childNodes[1].textContent;
    
    const firstContactIndex = tableContent.indexOf("Elephantress");
    const secondContactIndex = tableContent.indexOf("Jimbo");
    const thirdContactIndex = tableContent.indexOf("Smith");
    
    expect(screen.getByText("1 results").textContent).toBe("1 results");
    // Zebra Elephantress should not be in the table
    expect(firstContactIndex).toBe(-1);
    // Sleepy Jimbo should not be in the table
    expect(secondContactIndex).toBe(-1);
    // Adam Smith should be in the table due to his email matching keyword "westward"
    expect(thirdContactIndex).toBeGreaterThan(-1);
});

test("sort by field change should update state", () => {
    render(<ProviderList contacts={multipleValidContacts}/>);
    const sortByDropDown = screen.getByDisplayValue('Last Name');
    expect(sortByDropDown.value).toBe("lastName");

    fireEvent.change(sortByDropDown, { target: { value: "practiceName"}});

    expect(sortByDropDown.value).toBe("practiceName");
})

test("check on clicks trigger prop functions", () => {
    const mockOpenFormPanel = jest.fn();
    const mockDeleteProvider = jest.fn();

    render(<ProviderList contacts={multipleValidContacts} 
        openFormPanel={mockOpenFormPanel}
        deleteProvider={mockDeleteProvider} />);

    fireEvent.click(screen.getAllByText("Edit",  {selector: 'button'})[0]);
    fireEvent.click(screen.getAllByText("Delete")[0]);
    fireEvent.click(screen.getByText("âž• Add New Provider"));

    expect(mockOpenFormPanel).toBeCalledTimes(2);
    expect(mockDeleteProvider).toBeCalledTimes(1);

});