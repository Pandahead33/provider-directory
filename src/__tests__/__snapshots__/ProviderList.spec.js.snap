// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render and matches snapshot 1`] = `
<div
  className="provider-list"
>
  <div
    className="search-container"
  >
    <h1>
      Provider List
    </h1>
    <input
      className="add-new-provider-button"
      onClick={[Function]}
      type="button"
      value="➕ Add New Provider"
    />
    <input
      className="add-new-provider-button"
      type="button"
      value="☠️ Delete All Providers"
    />
    <div
      className="search-bar-container"
    >
      <input
        className="search"
        onChange={[Function]}
        placeholder="Search..."
        type="search"
      />
      <label
        htmlFor="sortBy"
      >
        <div
          className="sort-by-label"
        >
          Sort By
        </div>
        <select
          id="sortBy"
          name="sortBy"
          onChange={[Function]}
          value="lastName"
        >
          <option
            value="firstName"
          >
            First Name
          </option>
          <option
            value="lastName"
          >
            Last Name
          </option>
          <option
            value="emailAddress"
          >
            Email Address
          </option>
          <option
            value="speciality"
          >
            Speciality
          </option>
          <option
            value="practiceName"
          >
            Practice Name
          </option>
        </select>
      </label>
      <div
        className="sort-direction"
      >
        <div>
          <label
            htmlFor="ascending"
          >
            <input
              checked={true}
              id="ascending"
              name="sortDirection"
              onChange={[Function]}
              type="radio"
              value="ascending"
            />
            Ascending
          </label>
        </div>
        <div>
          <label
            htmlFor="descending"
          >
            <input
              checked={false}
              id="descending"
              name="sortDirection"
              onChange={[Function]}
              type="radio"
              value="descending"
            />
            Descending
          </label>
        </div>
      </div>
    </div>
  </div>
  <span
    className="results"
  >
    0
     results
  </span>
  <p>
    No providers found.
  </p>
</div>
`;
